### PR TITLE
mon: update mgr CLI command to display detailed output for disabled modules

### PR DESCRIPTION
**Pull request by Acewvrs**
Related Trello card: https://trello.com/c/eDEORDHm/595-ceph-mgr-module-ls-should-display-detailed-output-for-enabled-modules-like-it-does-for-disabled-modules

This PR enhances the ceph mgr module ls -f json-pretty command to include detailed information about “enabled” and “always-on” modules. Previously, the command only displayed detailed information for disabled modules, and this update ensures comprehensive JSON output for all module states.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Review by ljflores — COMMENTED** [review]
Please rebase the branch and fix the conflicts like we discussed during the RCOS meeting. Once you resolve conflicts, you should see the "needs-rebase" label get removed by the GitHub bot.

**Comment by Acewvrs** [comment]
> Please rebase the branch and fix the conflicts like we discussed during the RCOS meeting. Once you resolve conflicts, you should see the "needs-rebase" label get removed by the GitHub bot.

Hi, the merge conflict has been resolved, but a different error occurred after rebasing.

**Comment by ljflores** [comment]
Oh I see, the rebase label was removed. Good!

Let's see about the remaining test failures...
1. One of the commits still isn't signed off correctly (you can check the "details" tab in the "Signed-off-by" check.
2. The api check failed from a Python unit test. At a glance, it doesn't look related since your PR didn't make changes to the Python code. But you can inspect that further by checking the "details" tab. If you want to rerun the api check to try and get it to pass, you can do so by commenting "jenkins test api".
3. The make check test failed from a seemingly unrelated issue too:
```
[PostBuildScript] - [INFO] Executing post build scripts.
[PostBuildScript] - [INFO] Build does not have any of the results [ABORTED]. Did not execute build step #0.
INFO: Processing CTest-Version 3.x (default)
Agent went offline during the build
ERROR: Connection was broken
Build step 'Publish xUnit test result report' marked build as failure
```
Basically, if a test fails, you'll want to check if there is a compile time problem related to the part of the code that you changed. This seems like an infrastructure issue, where the node that was running the unit tests broke its connection. I would also retrigger this test to try to get it to pass with "jenkins test make check".

**Comment by ljflores** [comment]
Also, please add a description of the change to the pull request. You can click on the three dots above "Contribution Guidelines" to edit the description.

**Comment by Acewvrs** [comment]
jenkins test make check

**Comment by Acewvrs** [comment]
jenkins test api

**Comment by Acewvrs** [comment]
jenkins test api

**Comment by Acewvrs** [comment]
> Oh I see, the rebase label was removed. Good!
> 
> Let's see about the remaining test failures...
> 
> 1. One of the commits still isn't signed off correctly (you can check the "details" tab in the "Signed-off-by" check.
> 2. The api check failed from a Python unit test. At a glance, it doesn't look related since your PR didn't make changes to the Python code. But you can inspect that further by checking the "details" tab. If you want to rerun the api check to try and get it to pass, you can do so by commenting "jenkins test api".
> 3. The make check test failed from a seemingly unrelated issue too:
> 
> ```
> [PostBuildScript] - [INFO] Executing post build scripts.
> [PostBuildScript] - [INFO] Build does not have any of the results [ABORTED]. Did not execute build step #0.
> INFO: Processing CTest-Version 3.x (default)
> Agent went offline during the build
> ERROR: Connection was broken
> Build step 'Publish xUnit test result report' marked build as failure
> ```
> 
> Basically, if a test fails, you'll want to check if there is a compile time problem related to the part of the code that you changed. This seems like an infrastructure issue, where the node that was running the unit tests broke its connection. I would also retrigger this test to try to get it to pass with "jenkins test make check".

Thank you! I got the make check to pass. I'll rerun the API test now.

**Comment by Acewvrs** [comment]
jenkins test api

**Comment by Acewvrs** [comment]
jenkins test make check

**Comment by Acewvrs** [comment]
jenkins test api

**Comment by Acewvrs** [comment]
jenkins test api

**Comment by Acewvrs** [comment]
jenkins test api

**Comment by Acewvrs** [comment]
jenkins test api

**Comment by Acewvrs** [comment]
jenkins test make check

**Review by ljflores — APPROVED** [review]
LGTM. I am going to request a few more reviews before this goes into quality assurance testing to be sure.

File: src/mon/MgrMonitor.cc [hunk 1]
```diff
@@ -1012,13 +1012,24 @@ bool MgrMonitor::preprocess_command(MonOpRequestRef op)
       f->dump_object("mgrmap", m);
     }
     f->flush(rdata);
-  } else if (prefix == "mgr module ls") {
+} else if (prefix == "mgr module ls") {
```
Comment (ronen-fr) [inline src/mon/MgrMonitor.cc:1015]: why this indentation change?

File: src/mon/MgrMonitor.cc [hunk 2]
```diff
@@ -1012,13 +1012,24 @@ bool MgrMonitor::preprocess_command(MonOpRequestRef op)
...
     if (f) {
+      // Create a mapping from module names to ModuleInfo pointers
+      std::unordered_map<std::string, const MgrMap::ModuleInfo*> module_info_map;
```
Comment (ronen-fr) [inline src/mon/MgrMonitor.cc:1018]: You are manipulating pointers into a vector, if I'm not mistaken. I do not know the locking infra for the modules. Are we sure there is no possibility of change to the modules vector by another thread during the execution of this function?

File: src/mon/MgrMonitor.cc [hunk 3]
```diff
@@ -1012,13 +1012,24 @@ bool MgrMonitor::preprocess_command(MonOpRequestRef op)
...
+      const auto always_on_modules = map.get_always_on_modules();
+      for (const auto& mi : map.available_modules) {
+        module_info_map[mi.name] = &mi;
+      }
+
       f->open_object_section("modules");
       {
+        // always_on_modules
```
[the remark this answers is not in the document]
Comment (Acewvrs) [reply src/mon/MgrMonitor.cc:1026]: > If you need a block - better just name it (as an aux function)

Would it be okay if I just remove the block, instead of creating a whole new function?

File: src/mon/MgrMonitor.cc (hunk 2 above)
Comment (rzarzynski) [reply src/mon/MgrMonitor.cc]: Yeah. How about dropping the intermediary container and use e.g. `std::find_if` instead?

File: qa/tasks/mgr/mgr_test_case.py [hunk 4]
```diff
@@ -109,9 +109,7 @@ def setup_mgrs(cls):
         loaded = json.loads(cls.mgr_cluster.mon_manager.raw_cluster_cmd(
                    "mgr", "module", "ls", "--format=json-pretty"))['enabled_modules']
         module_names = {module['name'] for module in loaded}
-        print("module names: ", module_names)
         unload_modules = module_names - {"cephadm"}
-        print("unload modules: ", unload_modules)
```
Comment (rzarzynski) [inline qa/tasks/mgr/mgr_test_case.py]: Please drop this and previous commit – together they form _no change_.

**Review by rzarzynski — COMMENTED** [review]
Thank you for bringing this! Please see the comments below.

**Comment by Acewvrs** [comment]
jenkins test windows

**Comment by Jordan231111** [comment]
jenkins test make check arm64

**Comment by Acewvrs** [comment]
jenkins test make check

**Comment by Acewvrs** [comment]
jenkins test make check arm64

**Comment by Acewvrs** [comment]
jenkins test make check

**Comment by Acewvrs** [comment]
jenkins test make check arm64

**Comment by Acewvrs** [comment]
jenkins test make check

**Comment by Acewvrs** [comment]
jenkins test make check

**Review by ljflores — COMMENTED** [review]
Hi @Acewvrs, the logic looks good to me. Can you take one more step to clean up the commits and squash them into one or a few distinct logical changes? For instance, "mon: fix spacing" can be grouped in with another commit since this is an after-cleanup.

I have also re-requested the reviews of those who dropped comments before in case there are any final changes to be made. Then, this can go into QA!

**Comment by ljflores** [comment]
If you are unable to complete the PR, let me know and I can take it over.

**Comment by Acewvrs** [comment]
Hi @ljflores, I've squashed the commits and grouped them into 3 main categories:

1. Our initial approach to the problem using a hash map.
2. Fixing the test to accommodate the changes we made from group 1. 
3. Our more refined solution, which reduces memory usage by removing the hash map.

I've added brief descriptions to each commit as well.

File: src/mon/MgrMonitor.cc [hunk 5]
```diff
@@ -1014,43 +1014,60 @@ bool MgrMonitor::preprocess_command(MonOpRequestRef op)
...
       f->open_object_section("modules");
-      {
-        f->open_array_section("always_on_modules");
-        for (auto& p : map.get_always_on_modules()) {
-          f->dump_string("module", p);
-        }
-        f->close_section();
+      
+      // always_on_modules
+      f->open_array_section("always_on_modules");
```
Comment (ronen-fr) [inline src/mon/MgrMonitor.cc:1023]: (not a must for this PR, but if you feel like it:
you can change - in a separate commit - the way sections are opened and closed in the new code to
use the safer RAII wrappers. See for example https://github.com/ceph/ceph/pull/61618 for details).

File: src/mon/MgrMonitor.cc [hunk 6]
```diff
@@ -1014,43 +1014,60 @@ bool MgrMonitor::preprocess_command(MonOpRequestRef op)
     f->flush(rdata);
   } else if (prefix == "mgr module ls") {
     if (f) {
+      const auto always_on_modules = map.get_always_on_modules();
```
Comment (ronen-fr) [inline src/mon/MgrMonitor.cc:1017]: please don't reuse the exact name of the map member. It's confusing.

File: src/mon/MgrMonitor.cc [hunk 7]
```diff
@@ -1014,43 +1014,60 @@ bool MgrMonitor::preprocess_command(MonOpRequestRef op)
...
+      }
+      f->close_section();
+
+      // enabled_modules
+      f->open_array_section("enabled_modules");
+      for (const auto& module_name : map.modules) {
+        if (always_on_modules.count(module_name) > 0)
+          continue;
+        auto module_itr = std::find_if(
+          available_modules.begin(),
```
Comment (ronen-fr) [inline src/mon/MgrMonitor.cc:1048]: I am missing something...  isn't your available_modules the set that you have created earlier? a set of strings (the module names)? if so - why find_if(), and not just find()?

File: src/mon/MgrMonitor.cc [hunk 8]
```diff
@@ -1014,43 +1014,60 @@ bool MgrMonitor::preprocess_command(MonOpRequestRef op)
...
+        if (module_itr != available_modules.end()) {
+          (*module_itr).dump(f.get());
+        } 
+      }
+      f->close_section();
+
+      // disabled_modules
+      f->open_array_section("disabled_modules");
+      for (const auto& p : map.available_modules) {
+        if (map.modules.count(p.name) == 0 &&
```
Comment (ronen-fr) [inline src/mon/MgrMonitor.cc:1061]: contains

**Comment by Acewvrs** [comment]
jenkins test api

File: src/mon/MgrMonitor.cc [hunk 9]
```diff
@@ -1014,43 +1014,60 @@ bool MgrMonitor::preprocess_command(MonOpRequestRef op)
...
-        f->open_array_section("always_on_modules");
-        for (auto& p : map.get_always_on_modules()) {
-          f->dump_string("module", p);
-        }
-        f->close_section();
+      
+      // always_on_modules
+      f->open_array_section("always_on_modules");
+      for (const auto& module_name : always_on_modules) {
+        auto module_itr = std::find_if(
```
[the remark this answers is not in the document]
Comment (rzarzynski) [reply src/mon/MgrMonitor.cc:1025]: Would be nice!

File: src/mon/MgrMonitor.cc (hunk 7 above)
Comment (rzarzynski) [reply src/mon/MgrMonitor.cc:1048]: My understanding is that:

1. `available_modules` is as below and stores `MgrMap::ModuleInfo` instances:
```cpp
 const auto available_modules = map.available_modules;
```
2. The `find_if` is for the sake of accessing the `name` member of `MgrMap::ModuleInfo`.

**Comment by ljflores** [comment]
Hey @Acewvrs, IMO the 4 commits can be squashed into just one commit. Can you squash it, and consolidate the explanations into one description?

@ronen-fr Can you take a final look at this before it enters QA?

**Comment by Acewvrs** [comment]
> Hey @Acewvrs, IMO the 4 commits can be squashed into just one commit. Can you squash it, and consolidate the explanations into one description?

I’ve squashed the 4 commits into a single commit.

**Review by ronen-fr — APPROVED** [review]
LGTM.
One suggested improvement: use the RAII "dump sections" (see https://github.com/ceph/ceph/pull/61506). But this can be done in
a followup PR.

**Comment by Acewvrs** [comment]
jenkins test api

**Comment by Acewvrs** [comment]
jenkins test api

**Comment by Acewvrs** [comment]
jenkins test make check

**Comment by Acewvrs** [comment]
Jenkins test api

**Comment by Acewvrs** [comment]
Jenkins test api

**Comment by Acewvrs** [comment]
Jenkins test api

**Comment by Acewvrs** [comment]
Jenkins test api

**Comment by Acewvrs** [comment]
jenkins test api

**Comment by Acewvrs** [comment]
jenkins test api

**Comment by Acewvrs** [comment]
jenkins test api

**Comment by Acewvrs** [comment]
jenkins test api

**Comment by Acewvrs** [comment]
jenkins test api

**Comment by Acewvrs** [comment]
jenkins test api

**Comment by Acewvrs** [comment]
jenkins test api

**Comment by Acewvrs** [comment]
jenkins test api

**Comment by Acewvrs** [comment]
jenkins test api

**Comment by Acewvrs** [comment]
jenkins test api

**Comment by Acewvrs** [comment]
jenkins test api

**Comment by Acewvrs** [comment]
jenkins test api

**Comment by Acewvrs** [comment]
jenkins test api

**Comment by Acewvrs** [comment]
jenkins test api
